### PR TITLE
Numeric for loops!

### PR DESCRIPTION
**Pull request by AmaranthineCodices**
This is actually a bundle of inter-related changes. It adds support for parsing numeric for loops, and the various things that happened along the way. To summarize:

* NumericFor added to the AST
* AST tokens now derive `PartialEq`
* Tokenizer parses `for`, `do`, and `end` keywords
* Parser can parse identifiers as expressions
* Unit test added for function call parsing
* Unit test added for numeric for parsing
* A bunch of integration tests for this

### CODE REVIEW

File: src/ast.rs [hunk 1]
```diff
@@ -18,12 +18,21 @@ pub enum Expression<'a> {
...
     pub name: &'a str,
     pub value: Expression<'a>,
 }
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct NumericFor<'a> {
+    pub var: &'a str,
+    pub start: Expression<'a>,
+    pub end: Expression<'a>,
+    pub step: Expression<'a>,
```
Comment (LPGhatguy) [inline src/ast.rs]: Should `step` instead be an `Option<Expression<'a>>`, to indicate that the source might have not specified the step at all?

File: parse_examples/should_not_parse/for_loop_3.lua [hunk 2]
```diff
@@ -0,0 +1,3 @@
+for i = 1, 2, do
+    print(i)
+end
```
Comment (LPGhatguy) [inline parse_examples/should_not_parse/for_loop_3.lua:3]: Can we add a test case for a for loop with no body, as well?

File: src/ast.rs (hunk 1 above)
Comment (AmaranthineCodices) [reply src/ast.rs]: It should, yes! Fixed now.

**Comment by AmaranthineCodices** [comment]
Thought: Should `NumericFor`'s expression fields be boxed, like `FunctionCall`'s are?

File: parse_examples/should_not_parse/for_loop_3.lua (hunk 2 above)
Comment (AmaranthineCodices) [reply parse_examples/should_not_parse/for_loop_3.lua:3]: Done!

**Comment by LPGhatguy** [comment]
I'm not sure the best place to introduce `Box` into data structures like this, but I imagine we should be consistent.

Yeah, let's box every `Expression` and `Statement`!

**Comment by LPGhatguy** [comment]
I accidentally made a conflict -- I'll resolve, merge, and then work on this in master!

**Pull request merged**